cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -411,6 +432,30 @@ public boolean updateAppAllowed(ApplicationId appId) throws Exception {
     return version.endsWith(ApplicationId.DEFAULT_VERSION);
   }
 
+  /**
+   * To determine whether the app is allowed to be deployed:
+   * Do not deploy when the parent version is not the latest.
+   *
+   * @param appId the id of the application to be determined
+   * @param parentVersion the version of the application from which the app is deployed
+   * @return whether the app version is allowed to be deployed
+   */
+  public boolean deployAppAllowed(ApplicationId appId, @Nullable String parentVersion)  {
```

Comment:
we should not have this method, as it encourages duplicate DB calls and is open to race conditions. Whatever is calling this is also likely getting the latest app somewhere else. 

More importantly, the parent version check needs to happen in the same transaction that adds the new row, otherwise two requests that happen at the same time can end up writing different versions.

Reply:
for the deployment to be transactional - we will update the `latest` column in the table. We don't have to check for `parentVersion`